File: typescript/src/judge.ts
/**
 * LLM judge interface for AI-judged graders.
 *
 * Provides a pluggable judge abstraction for subjective quality assessment.
 * See SPEC.md §10 (Pluggable Judges).
 */

import type { Grade } from "./types.js";

/** Base interface for LLM-based judges. See SPEC §10. */
export interface Judge {
  /** Evaluate a subject answer against a criterion. */
  evaluate(
    criterion: string,
    subjectAnswer: string,
    context?: Record<string, unknown>,
  ): Grade | Promise<Grade>;
}

/** A no-op judge that returns a skipped grade. Used when no judge is configured. */
export class NullJudge implements Judge {
  evaluate(
    criterion: string,
    _subjectAnswer: string,
    _context?: Record<string, unknown>,
  ): Grade {
    return {
      criterion,
      score: 0,
      metric: "quality",
      passed: false,
      detail: "No judge configured.",
      layer: "ai_judged",
      skipped: true,
    };
  }
}

File: go/judge.go
package beval

// Judge is the interface for LLM-based judges. See SPEC §10.
type Judge interface {
	// Evaluate assesses a subject answer against a criterion.
	// Returns a Grade with score, metric, and reasoning detail.
	Evaluate(criterion string, subjectAnswer string, context map[string]interface{}) Grade
}

// NullJudge is a no-op judge that returns a skipped grade.
// Used when no judge is configured.
type NullJudge struct{}

// Evaluate returns a skipped grade indicating no judge is configured.
func (n NullJudge) Evaluate(criterion string, subjectAnswer string, context map[string]interface{}) Grade {
	return Grade{
		Criterion: criterion,
		Score:     0.0,
		Metric:    "quality",
		Passed:    false,
		Detail:    "No judge configured.",
		Layer:     LayerAIJudged,
		Skipped:   true,
	}
}
